File: cli/internal/wikis/checks.go
package wikis

import (
	"fmt"
	"math"
	"path"
	"sort"
	"strings"
)

const (
	CheckIndexMD     = "check-index-md"
	CheckTitles      = "check-titles"
	CheckFrontmatter = "check-frontmatter"
	CheckAliases     = "check-aliases"
	CheckCovers      = "check-covers"
	CheckLinks       = "check-links"
)

var AllCheckNames = []string{CheckIndexMD, CheckTitles, CheckFrontmatter, CheckAliases, CheckCovers, CheckLinks}

func RunCheck(root, name string, opts CheckOptions) (CheckResult, error) {
	corpusOpts := LoadCorpusOptions{}
	if opts.FailFast {
		switch name {
		case CheckTitles:
			corpusOpts.FailFastParse = true
			corpusOpts.ParseCodes = parseCodeSet("missing_h1", "multiple_h1")
		case CheckFrontmatter:
			corpusOpts.FailFastParse = true
			corpusOpts.ParseCodes = parseCodeSet("unknown_frontmatter_field", "invalid_frontmatter_type", "invalid_covers_type", "invalid_intents_type")
		}
	}
	corpus, parseErrs, err := LoadCorpusWithOptions(root, corpusOpts)
	if err != nil {
		return CheckResult{}, err
	}
	var errs []CheckError
	switch name {
	case CheckIndexMD:
		errs = checkIndexMD(corpus, opts)
	case CheckTitles:
		errs = filterParseErrs(parseErrs, "missing_h1", "multiple_h1")
	case CheckFrontmatter:
		errs = filterParseErrs(parseErrs, "unknown_frontmatter_field", "invalid_frontmatter_type", "invalid_covers_type")
		if !(opts.FailFast && len(errs) > 0) {
			errs = append(errs, checkFrontmatter(corpus, opts)...)
		}
	case CheckAliases:
		errs = checkAliases(corpus, opts)
	case CheckCovers:
		errs = checkCovers(corpus, opts)
	case CheckLinks:
		errs = checkLinks(corpus, opts)
	default:
		return CheckResult{}, fmt.Errorf("unknown wikis check: %s", name)
	}
	if opts.FailFast && len(errs) > 1 {
		errs = errs[:1]
	}
	for i := range errs {
		errs[i].Check = name
	}
	return makeCheckResult(name, errs, opts), nil
}

func RunAllChecks(root string, opts CheckOptions) ([]CheckResult, error) {
	var results []CheckResult
	rawOpts := opts
	rawOpts.MaxErrors = math.MaxInt
	for _, name := range AllCheckNames {
		result, err := RunCheck(root, name, rawOpts)
		if err != nil {
			return nil, err
		}
		results = append(results, result)
		if opts.FailFast && result.TotalErrors > 0 {
			break
		}
	}
	return trimCheckAllResults(results, opts), nil
}

func makeCheckResult(name string, errs []CheckError, opts CheckOptions) CheckResult {
	limit := opts.MaxErrors
	if limit <= 0 {
		limit = 100
	}
	shown := errs
	if len(shown) > limit {
		shown = shown[:limit]
	}
	if shown == nil {
		shown = []CheckError{}
	}
	return CheckResult{
		Check:        name,
		OK:           len(errs) == 0,
		TotalErrors:  len(errs),
		ShownErrors:  len(shown),
		HiddenErrors: len(errs) - len(shown),
		Truncated:    len(errs) > len(shown),
		Errors:       shown,
	}
}

func trimCheckAllResults(results []CheckResult, opts CheckOptions) []CheckResult {
	remaining := opts.MaxErrors
	if remaining <= 0 {
		remaining = 500
	}
	for i := range results {
		errors := results[i].Errors
		if len(errors) > remaining {
			errors = errors[:remaining]
		}
		if errors == nil {
			errors = []CheckError{}
		}
		remaining -= len(errors)
		if remaining < 0 {
			remaining = 0
		}
		results[i].Errors = errors
		results[i].ShownErrors = len(errors)
		results[i].HiddenErrors = results[i].TotalErrors - len(errors)
		results[i].Truncated = results[i].HiddenErrors > 0
	}
	return results
}

func filterParseErrs(errs []CheckError, codes ...string) []CheckError {
	allowed := map[string]bool{}
	for _, code := range codes {
		allowed[code] = true
	}
	var out []CheckError
	for _, err := range errs {
		if allowed[err.Code] {
			out = append(out, err)
		}
	}
	return out
}

func parseCodeSet(codes ...string) map[string]bool {
	allowed := map[string]bool{}
	for _, code := range codes {
		allowed[code] = true
	}
	return allowed
}

func checkIndexMD(c Corpus, opts CheckOptions) []CheckError {
	dirs := map[string]bool{}
	for _, doc := range c.Docs {
		if strings.HasPrefix(doc.Path, "routing/") {
			continue
		}
		dir := path.Dir(doc.Path)
		for dir != "." && dir != "" {
			dirs[dir] = true
			parent := path.Dir(dir)
			if parent == dir {
				break
			}
			dir = parent
		}
	}
	var keys []string
	for dir := range dirs {
		keys = append(keys, dir)
	}
	sort.Strings(keys)
	var errs []CheckError
	for _, dir := range keys {
		indexPath := dir + "/index.md"
		if c.ByPath[indexPath] == nil {
			errs = append(errs, CheckError{Path: dir, Code: "missing_index_md", Message: "directory is missing index.md", Target: indexPath})
			if opts.FailFast {
				return errs
			}
		}
	}
	return errs
}

func checkFrontmatter(c Corpus, opts CheckOptions) []CheckError {
	var errs []CheckError
	add := func(err CheckError) bool {
		errs = append(errs, err)
		return opts.FailFast
	}
	for _, doc := range c.Docs {
		switch {
		case doc.Kind == KindSpec && doc.SpecType == SpecTypeMetric:
			if !doc.HasFrontmatter {
				if add(CheckError{Path: doc.Path, Code: "missing_frontmatter", Message: "metric spec must have frontmatter"}) {
					return errs
				}
				continue
			}
			if doc.Name == "" {
				if add(CheckError{Path: doc.Path, Code: "missing_name", Message: "metric spec must have name", Target: "name"}) {
					return errs
				}
			}
			if doc.Label == "" {
				if add(CheckError{Path: doc.Path, Code: "missing_label", Message: "metric spec must have label", Target: "label"}) {
					return errs
				}
			}
		case doc.Kind == KindPlaybook && doc.Playbook.IsCombo:
			if !doc.HasFrontmatter {
				if add(CheckError{Path: doc.Path, Code: "missing_frontmatter", Message: "combo playbook must have frontmatter"}) {
					return errs
				}
				continue
			}
			if doc.Aliases == nil {
				if add(CheckError{Path: doc.Path, Code: "missing_required_field", Message: "combo playbook must have aliases", Target: "aliases"}) {
					return errs
				}
			}
			if doc.Covers == nil {
				if add(CheckError{Path: doc.Path, Code: "missing_covers", Message: "combo playbook must have covers", Target: "covers"}) {
					return errs
				}
			}
		}
		if len(doc.Playbook.Intents) > 0 {
			if doc.Kind != KindPlaybook || !doc.Playbook.IsSingle {
				if add(CheckError{Path: doc.Path, Code: "invalid_intents_target", Message: "intents are only allowed on single metric playbooks", Target: "intents"}) {
					return errs
				}
			}
			for intentName, intent := range doc.Playbook.Intents {
				if strings.TrimSpace(intentName) == "" {
					if add(CheckError{Path: doc.Path, Code: "invalid_intents_type", Message: "intent name must not be empty", Target: "intents"}) {
						return errs
					}
				}
				if len(intent.Aliases) == 0 {
					if add(CheckError{Path: doc.Path, Code: "invalid_intents_type", Message: "intent aliases must not be empty", Target: "intents." + intentName + ".aliases"}) {
						return errs
					}
				}
				seen := map[string]bool{}
				for _, alias := range intent.Aliases {
					if strings.TrimSpace(alias) == "" {
						if add(CheckError{Path: doc.Path, Code: "invalid_intents_type", Message: "intent alias must not be empty", Target: "intents." + intentName + ".aliases"}) {
							return errs
						}
					}
					if seen[alias] {
						if add(CheckError{Path: doc.Path, Code: "duplicate_intent_alias", Message: "duplicate intent alias in playbook intent", Target: "intents." + intentName + ".aliases", Value: alias}) {
							return errs
						}
					}
					seen[alias] = true
				}
			}
		}
	}
	return errs
}

func checkAliases(c Corpus, opts CheckOptions) []CheckError {
	var errs []CheckError
	add := func(err CheckError) bool {
		errs = append(errs, err)
		return opts.FailFast
	}
	global := map[string]CheckError{}
	for _, doc := range c.Docs {
		if doc.Kind == KindSpec && doc.SpecType == SpecTypeMetric {
			if doc.Name == "" {
				if add(CheckError{Path: doc.Path, Code: "missing_name", Message: "metric spec must have name", Target: "name"}) {
					return errs
				}
			}
			if doc.Label == "" {
				if add(CheckError{Path: doc.Path, Code: "missing_label", Message: "metric spec must have label", Target: "label"}) {
					return errs
				}
			}
		}
		if doc.Kind == KindPlaybook && doc.Playbook.IsSingle && len(doc.Aliases) > 0 {
			if add(CheckError{Path: doc.Path, Code: "alias_not_allowed", Message: "single metric playbook must not maintain aliases", Target: "aliases"}) {
				return errs
			}
		}
		if doc.Kind == KindPlaybook && doc.Playbook.IsCombo && len(doc.Aliases) == 0 {
			if add(CheckError{Path: doc.Path, Code: "missing_required_field", Message: "combo playbook must maintain aliases", Target: "aliases"}) {
				return errs
			}
		}
		seen := map[string]bool{}
		for _, alias := range doc.Aliases {
			if seen[alias] {
				if add(CheckError{Path: doc.Path, Code: "duplicate_alias", Message: "duplicate alias in document", Value: alias}) {
					return errs
				}
			}
			seen[alias] = true
		}
		for _, recall := range orderedRecallValues(doc) {
			for _, value := range recall.Values {
				if value == "" {
					continue
				}
				if other, ok := global[value]; ok {
					if add(CheckError{Path: doc.Path, Code: "duplicate_recall_value", Message: "duplicate recall value", Target: recall.Field, Value: value, Other: other.Path}) {
						return errs
					}
				} else {
					global[value] = CheckError{Path: doc.Path, Target: recall.Field}
				}
			}
		}
	}
	return errs
}

type recallFieldValues struct {
	Field  string
	Values []string
}

func orderedRecallValues(doc Document) []recallFieldValues {
	var out []recallFieldValues
	if doc.Kind == KindSpec {
		if doc.SpecType == SpecTypeMetric || doc.Name != "" || doc.Label != "" || len(doc.Aliases) > 0 {
			out = append(out,
				recallFieldValues{Field: "name", Values: []string{doc.Name}},
				recallFieldValues{Field: "label", Values: []string{doc.Label}},
				recallFieldValues{Field: "aliases", Values: doc.Aliases},
			)
		}
	}
	if doc.Kind == KindPlaybook && doc.Playbook.IsCombo {
		out = append(out, recallFieldValues{Field: "aliases", Values: doc.Aliases})
	}
	return out
}

func recallValues(doc Document) map[string][]string {
	out := map[string][]string{}
	if doc.Kind == KindSpec {
		if doc.SpecType == SpecTypeMetric || doc.Name != "" || doc.Label != "" || len(doc.Aliases) > 0 {
			out["name"] = []string{doc.Name}
			out["label"] = []string{doc.Label}
			out["aliases"] = doc.Aliases
		}
	}
	if doc.Kind == KindPlaybook && doc.Playbook.IsCombo {
		out["aliases"] = doc.Aliases
	}
	return out
}

func checkCovers(c Corpus, opts CheckOptions) []CheckError {
	var errs []CheckError
	add := func(err CheckError) bool {
		errs = append(errs, err)
		return opts.FailFast
	}
	for _, doc := range c.Docs {
		if doc.Kind != KindPlaybook || !doc.Playbook.IsCombo {
			continue
		}
		if doc.Covers == nil {
			if add(CheckError{Path: doc.Path, Code: "missing_covers", Message: "combo playbook must maintain covers", Target: "covers"}) {
				return errs
			}
			continue
		}
		for _, cover := range doc.Covers {
			if !isSpecDocPath(cover) || !strings.HasSuffix(cover, ".md") {
				if add(CheckError{Path: doc.Path, Code: "invalid_cover_path", Message: "cover must reference a spec logical path", Value: cover}) {
					return errs
				}
				continue
			}
			if !c.SpecPaths[cover] {
				if add(CheckError{Path: doc.Path, Code: "missing_cover_target", Message: "cover target does not exist", Value: cover}) {
					return errs
				}
			}
		}
	}
	return errs
}

func checkLinks(c Corpus, opts CheckOptions) []CheckError {
	var errs []CheckError
	add := func(err CheckError) bool {
		errs = append(errs, err)
		return opts.FailFast
	}
	for _, doc := range c.Docs {
		switch {
		case doc.Kind == KindSpec && doc.SpecType == SpecTypeMetric:
			if IsReferenceSpecPath(doc.Path) {
				continue
			}
			playbookPath := SamePath(doc.Path, "playbooks")
			if c.ByPath[playbookPath] == nil {
				if add(CheckError{Path: doc.Path, Code: "missing_playbook", Message: "metric spec is missing same-path playbook", Target: playbookPath}) {
					return errs
				}
			}
		case doc.Kind == KindTemplate && !doc.IsIndex:
			if c.ByPath[doc.Template.PlaybookPath] == nil {
				if add(CheckError{Path: doc.Path, Code: "orphan_template", Message: "template is missing same-path playbook", Target: doc.Template.PlaybookPath}) {
					return errs
				}
			}
		}
	}
	return errs
}

func IsReferenceSpecPath(logical string) bool {
	if strings.HasPrefix(logical, "dims/") || strings.HasPrefix(logical, "rules/") {
		return path.Base(logical) == "spec.md"
	}
	if !strings.HasPrefix(logical, "spec/") {
		return false
	}
	rest := strings.TrimPrefix(logical, "spec/")
	if strings.HasPrefix(rest, "common/") {
		return true
	}
	return strings.HasPrefix(rest, "dim-")
}

File: packages/data-harness-cli/src/lib/wikis/checks.js
import path from "node:path";

import { isSpecDocPath, loadCorpusWithOptions } from "./parse.js";
import { isReferenceSpecPath, KIND_PLAYBOOK, KIND_SPEC, KIND_TEMPLATE, samePath, SPEC_TYPE_METRIC } from "./paths.js";

export const CHECK_INDEX_MD = "check-index-md";
export const CHECK_TITLES = "check-titles";
export const CHECK_FRONTMATTER = "check-frontmatter";
export const CHECK_ALIASES = "check-aliases";
export const CHECK_COVERS = "check-covers";
export const CHECK_LINKS = "check-links";

export const ALL_CHECK_NAMES = [
  CHECK_INDEX_MD,
  CHECK_TITLES,
  CHECK_FRONTMATTER,
  CHECK_ALIASES,
  CHECK_COVERS,
  CHECK_LINKS,
];

export function runCheck(root, name, opts = {}) {
  const corpusOpts = {};
  if (opts.failFast) {
    if (name === CHECK_TITLES) {
      corpusOpts.failFastParse = true;
      corpusOpts.parseCodes = new Set(["missing_h1", "multiple_h1"]);
    } else if (name === CHECK_FRONTMATTER) {
      corpusOpts.failFastParse = true;
      corpusOpts.parseCodes = new Set([
        "unknown_frontmatter_field",
        "invalid_frontmatter_type",
        "invalid_covers_type",
        "invalid_intents_type",
      ]);
    }
  }
  const { corpus, errs: parseErrs } = loadCorpusWithOptions(root, corpusOpts);
  let errs = [];
  switch (name) {
    case CHECK_INDEX_MD:
      errs = checkIndexMD(corpus, opts);
      break;
    case CHECK_TITLES:
      errs = filterParseErrs(parseErrs, ["missing_h1", "multiple_h1"]);
      break;
    case CHECK_FRONTMATTER:
      errs = filterParseErrs(parseErrs, ["unknown_frontmatter_field", "invalid_frontmatter_type", "invalid_covers_type"]);
      if (!(opts.failFast && errs.length > 0)) {
        errs = [...errs, ...checkFrontmatter(corpus, opts)];
      }
      break;
    case CHECK_ALIASES:
      errs = checkAliases(corpus, opts);
      break;
    case CHECK_COVERS:
      errs = checkCovers(corpus, opts);
      break;
    case CHECK_LINKS:
      errs = checkLinks(corpus, opts);
      break;
    default:
      throw new Error(`unknown wikis check: ${name}`);
  }
  if (opts.failFast && errs.length > 1) errs = errs.slice(0, 1);
  for (const err of errs) err.check = name;
  return makeCheckResult(name, errs, opts);
}

export function runAllChecks(root, opts = {}) {
  const results = [];
  const rawOpts = { ...opts, maxErrors: Number.MAX_SAFE_INTEGER };
  for (const name of ALL_CHECK_NAMES) {
    const result = runCheck(root, name, rawOpts);
    results.push(result);
    if (opts.failFast && result.totalErrors > 0) break;
  }
  return trimCheckAllResults(results, opts);
}

export function makeCheckResult(name, errs, opts = {}) {
  let limit = opts.maxErrors || 0;
  if (limit <= 0) limit = 100;
  let shown = errs;
  if (shown.length > limit) shown = shown.slice(0, limit);
  return {
    check: name,
    ok: errs.length === 0,
    totalErrors: errs.length,
    shownErrors: shown.length,
    hiddenErrors: errs.length - shown.length,
    truncated: errs.length > shown.length,
    errors: shown,
  };
}

function trimCheckAllResults(results, opts = {}) {
  let remaining = opts.maxErrors || 0;
  if (remaining <= 0) remaining = 500;
  for (const result of results) {
    let errors = result.errors;
    if (errors.length > remaining) errors = errors.slice(0, remaining);
    remaining -= errors.length;
    if (remaining < 0) remaining = 0;
    result.errors = errors;
    result.shownErrors = errors.length;
    result.hiddenErrors = result.totalErrors - errors.length;
    result.truncated = result.hiddenErrors > 0;
  }
  return results;
}

function filterParseErrs(errs, codes) {
  const allowed = new Set(codes);
  return errs.filter((err) => allowed.has(err.code));
}

function checkIndexMD(c, opts) {
  const dirs = new Set();
  for (const doc of c.docs) {
    if (doc.path.startsWith("routing/")) continue;
    let dir = path.posix.dirname(doc.path);
    while (dir !== "." && dir !== "") {
      dirs.add(dir);
      const parent = path.posix.dirname(dir);
      if (parent === dir) break;
      dir = parent;
    }
  }
  const keys = [...dirs].sort();
  const errs = [];
  for (const dir of keys) {
    const indexPath = `${dir}/index.md`;
    if (!c.byPath[indexPath]) {
      errs.push({ path: dir, code: "missing_index_md", message: "directory is missing index.md", target: indexPath });
      if (opts.failFast) return errs;
    }
  }
  return errs;
}

function checkFrontmatter(c, opts) {
  const errs = [];
  const add = (err) => {
    errs.push(err);
    return opts.failFast;
  };
  for (const doc of c.docs) {
    if (doc.kind === KIND_SPEC && doc.specType === SPEC_TYPE_METRIC) {
      if (!doc.hasFrontmatter) {
        if (add({ path: doc.path, code: "missing_frontmatter", message: "metric spec must have frontmatter" })) return errs;
        continue;
      }
      if (!doc.name) {
        if (add({ path: doc.path, code: "missing_name", message: "metric spec must have name", target: "name" })) return errs;
      }
      if (!doc.label) {
        if (add({ path: doc.path, code: "missing_label", message: "metric spec must have label", target: "label" })) return errs;
      }
    } else if (doc.kind === KIND_PLAYBOOK && doc.playbook?.isCombo) {
      if (!doc.hasFrontmatter) {
        if (add({ path: doc.path, code: "missing_frontmatter", message: "combo playbook must have frontmatter" })) return errs;
        continue;
      }
      if (!doc.aliases) {
        if (add({ path: doc.path, code: "missing_required_field", message: "combo playbook must have aliases", target: "aliases" })) return errs;
      }
      if (!doc.covers) {
        if (add({ path: doc.path, code: "missing_covers", message: "combo playbook must have covers", target: "covers" })) return errs;
      }
    }
    const intents = doc.playbook?.intents;
    if (intents && Object.keys(intents).length > 0) {
      if (doc.kind !== KIND_PLAYBOOK || !doc.playbook?.isSingle) {
        if (add({ path: doc.path, code: "invalid_intents_target", message: "intents are only allowed on single metric playbooks", target: "intents" })) {
          return errs;
        }
      }
      for (const [intentName, intent] of Object.entries(intents)) {
        if (!String(intentName).trim()) {
          if (add({ path: doc.path, code: "invalid_intents_type", message: "intent name must not be empty", target: "intents" })) return errs;
        }
        if (!intent?.aliases?.length) {
          if (add({ path: doc.path, code: "invalid_intents_type", message: "intent aliases must not be empty", target: `intents.${intentName}.aliases` })) {
            return errs;
          }
        }
        const seen = new Set();
        for (const alias of intent.aliases || []) {
          if (!String(alias).trim()) {
            if (add({ path: doc.path, code: "invalid_intents_type", message: "intent alias must not be empty", target: `intents.${intentName}.aliases` })) {
              return errs;
            }
          }
          if (seen.has(alias)) {
            if (add({ path: doc.path, code: "duplicate_intent_alias", message: "duplicate intent alias in playbook intent", target: `intents.${intentName}.aliases`, value: alias })) {
              return errs;
            }
          }
          seen.add(alias);
        }
      }
    }
  }
  return errs;
}

function checkAliases(c, opts) {
  const errs = [];
  const add = (err) => {
    errs.push(err);
    return opts.failFast;
  };
  const global = new Map();
  for (const doc of c.docs) {
    if (doc.kind === KIND_SPEC && doc.specType === SPEC_TYPE_METRIC) {
      if (!doc.name) {
        if (add({ path: doc.path, code: "missing_name", message: "metric spec must have name", target: "name" })) return errs;
      }
      if (!doc.label) {
        if (add({ path: doc.path, code: "missing_label", message: "metric spec must have label", target: "label" })) return errs;
      }
    }
    if (doc.kind === KIND_PLAYBOOK && doc.playbook?.isSingle && (doc.aliases || []).length > 0) {
      if (add({ path: doc.path, code: "alias_not_allowed", message: "single metric playbook must not maintain aliases", target: "aliases" })) {
        return errs;
      }
    }
    if (doc.kind === KIND_PLAYBOOK && doc.playbook?.isCombo && (doc.aliases || []).length === 0) {
      if (add({ path: doc.path, code: "missing_required_field", message: "combo playbook must maintain aliases", target: "aliases" })) {
        return errs;
      }
    }
    const seen = new Set();
    for (const alias of doc.aliases || []) {
      if (seen.has(alias)) {
        if (add({ path: doc.path, code: "duplicate_alias", message: "duplicate alias in document", value: alias })) return errs;
      }
      seen.add(alias);
    }
    for (const recall of orderedRecallValues(doc)) {
      for (const value of recall.values) {
        if (!value) continue;
        const other = global.get(value);
        if (other) {
          if (add({ path: doc.path, code: "duplicate_recall_value", message: "duplicate recall value", target: recall.field, value, other: other.path })) {
            return errs;
          }
        } else {
          global.set(value, { path: doc.path, target: recall.field });
        }
      }
    }
  }
  return errs;
}

function orderedRecallValues(doc) {
  const out = [];
  if (doc.kind === KIND_SPEC) {
    if (doc.specType === SPEC_TYPE_METRIC || doc.name || doc.label || (doc.aliases || []).length > 0) {
      out.push({ field: "name", values: [doc.name || ""] });
      out.push({ field: "label", values: [doc.label || ""] });
      out.push({ field: "aliases", values: doc.aliases || [] });
    }
  }
  if (doc.kind === KIND_PLAYBOOK && doc.playbook?.isCombo) {
    out.push({ field: "aliases", values: doc.aliases || [] });
  }
  return out;
}

function checkCovers(c, opts) {
  const errs = [];
  const add = (err) => {
    errs.push(err);
    return opts.failFast;
  };
  for (const doc of c.docs) {
    if (doc.kind !== KIND_PLAYBOOK || !doc.playbook?.isCombo) continue;
    if (!doc.covers) {
      if (add({ path: doc.path, code: "missing_covers", message: "combo playbook must maintain covers", target: "covers" })) return errs;
      continue;
    }
    for (const cover of doc.covers) {
      if (!isSpecDocPath(cover) || !cover.endsWith(".md")) {
        if (add({ path: doc.path, code: "invalid_cover_path", message: "cover must reference a spec logical path", value: cover })) return errs;
        continue;
      }
      if (!c.specPaths[cover]) {
        if (add({ path: doc.path, code: "missing_cover_target", message: "cover target does not exist", value: cover })) return errs;
      }
    }
  }
  return errs;
}

function checkLinks(c, opts) {
  const errs = [];
  const add = (err) => {
    errs.push(err);
    return opts.failFast;
  };
  for (const doc of c.docs) {
    if (doc.kind === KIND_SPEC && doc.specType === SPEC_TYPE_METRIC) {
      if (isReferenceSpecPath(doc.path)) continue;
      const playbookPath = samePath(doc.path, "playbooks");
      if (!c.byPath[playbookPath]) {
        if (add({ path: doc.path, code: "missing_playbook", message: "metric spec is missing same-path playbook", target: playbookPath })) {
          return errs;
        }
      }
    } else if (doc.kind === KIND_TEMPLATE && !doc.isIndex) {
      if (!c.byPath[doc.template?.playbookPath]) {
        if (add({ path: doc.path, code: "orphan_template", message: "template is missing same-path playbook", target: doc.template?.playbookPath })) {
          return errs;
        }
      }
    }
  }
  return errs;
}


